middlewares/statsd.go

```diff
@@ -0,0 +1,85 @@
+package middlewares
+
+import (
+	"time"
+
+	"github.com/containous/traefik/log"
+	"github.com/containous/traefik/safe"
+	"github.com/containous/traefik/types"
+	kitlog "github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/metrics"
+	"github.com/go-kit/kit/metrics/statsd"
+)
+
+var _ Metrics = (Metrics)(nil)
+
+var statsdClient = statsd.New("traefik.", kitlog.LoggerFunc(func(keyvals ...interface{}) error {
+	log.Info(keyvals)
+	return nil
+}))
+var statsdTicker *time.Ticker
+
+// Statsd is an Implementation for Metrics that exposes statsd metrics for the latency
+// and the number of requests partitioned by status code and method.
+// - number of requests partitioned by status code and method
+// - request durations
+// - amount of retries happened
+type Statsd struct {
+	reqsCounter          metrics.Counter
+	reqDurationHistogram metrics.Histogram
+	retryCounter         metrics.Counter
+}
+
+func (s *Statsd) getReqsCounter() metrics.Counter {
+	return s.reqsCounter
+}
+
+func (s *Statsd) getReqDurationHistogram() metrics.Histogram {
+	return s.reqDurationHistogram
+}
+
+func (s *Statsd) getRetryCounter() metrics.Counter {
+	return s.retryCounter
+}
+
+// NewStatsD creates new instance of StatsD
+func NewStatsD(name string) *Statsd {
+	var m Statsd
+
+	m.reqsCounter = statsdClient.NewCounter(ddMetricsReqsName, 1.0).With("service", name)
+	m.reqDurationHistogram = statsdClient.NewTiming(ddMetricsLatencyName, 1.0).With("service", name)
+
+	return &m
+}
+
+// InitStatsdClient initializes metrics pusher and creates a statsdClient if not created already
+func InitStatsdClient(config *types.Statsd) *time.Ticker {
```

Comment:
While I thought about trying to generalize and factorize the code out, like @ldez mentioned, while the 2 files look very similar in pattern, they internally use different go-kit/metrics structures to initialize metrics, name them and return outbound structures, so there isn't much that can actually be shared, besides the logical steps of operation, which are similar due to the similarity in DataDog being an enhancement on top of StatsD protocol.